metric-anomaly-detector/src/main/java/org/hypertrace/alert/engine/metric/anomaly/detector/evaluator/BaselineRuleEvaluator.java

```diff
@@ -0,0 +1,184 @@
+  }
+
+  Optional<NotificationEvent> evaluateRule(
+      MetricAnomalyEventCondition metricAnomalyEventCondition, AlertTask alertTask)
+      throws IOException {
+    DynamicThresholdCondition dynamicThresholdCondition =
+        metricAnomalyEventCondition
+            .getViolationConditionList()
+            .get(0)
+            .getDynamicThresholdCondition();
+
+    Duration d = java.time.Duration.parse(dynamicThresholdCondition.getBaselineDuration());
+    long durationMillis = d.get(ChronoUnit.MILLIS);
+
+    // this query will fetch metric data, which will used for baseline calculation and evaluation
+    QueryRequest queryRequest =
+        queryRequestHandler.getQueryRequest(
+            metricAnomalyEventCondition,
+            alertTask.getTenantId(),
+            alertTask.getCurrentExecutionTime() - durationMillis,
+            alertTask.getCurrentExecutionTime());
+
+    LOGGER.debug("Query request {}", queryRequest);
+
+    Iterator<ResultSetChunk> iterator =
+        queryRequestHandler.executeQuery(
+            Map.of(TENANT_ID_KEY, alertTask.getTenantId()), queryRequest);
+
+    List<Double> metricValuesForBaseline = new ArrayList<>();
+    List<Double> metricValuesForEvaluation = new ArrayList<>();
+
+    iterator.forEachRemaining(
+        v -> {
+          for (Row row : v.getRowList()) {
+            if (row.getColumnCount() > 2
+                && row.getColumn(1).getValueType()
+                    == org.hypertrace.core.query.service.api.ValueType.STRING) {
+              double value = Double.parseDouble(row.getColumn(1).getString());
+              metricValuesForBaseline.add(value);
+              if (Long.parseLong(row.getColumn(0).getString())
+                  >= alertTask.getLastExecutionTime()) {
+                metricValuesForEvaluation.add(value);
+              }
+            }
+          }
+        });
+
+    Baseline baseline =
+        getBaseline(metricValuesForBaseline.stream().mapToDouble(Double::doubleValue).toArray());
+
+    int dataCount = 0, violationCount = 0;
+    for (Double metricValue : metricValuesForBaseline) {
+      dataCount++;
+      if (metricValue < baseline.getLowerBound().getDouble()
+          || metricValue > baseline.getUpperBound().getDouble()) {
+        violationCount++;
+      }
+    }
+
+    if (!StaticRuleEvaluator.isViolation(dataCount, violationCount)) {
+      return Optional.empty();
+    }
+
+    return getNotificationEvent(
+        alertTask,
+        dataCount,
+        violationCount,
+        metricValuesForEvaluation,
+        baseline.getLowerBound().getDouble(),
+        baseline.getUpperBound().getDouble());
+  }
+
+  private static Baseline getBaseline(double[] metricValueArray) {
+    Median median = new Median();
+    StandardDeviation standardDeviation = new StandardDeviation();
+    double medianValue = median.evaluate(metricValueArray);
+    double sd = standardDeviation.evaluate(metricValueArray);
+    double lowerBound = medianValue - (2 * sd);
```

Comment:
Nit : lowerBound = max(0,medianValue - (2*sd)) ?

Reply:
lowerBound can be -ve as well

Reply:
Yes so max with 0 ?